File: dotnet/Gherkin.Specs/Tokens/TestTokenFormatter.cs
namespace Gherkin.Specs.Tokens;

class TestTokenFormatter
{
    public string FormatToken(Token token)
    {
        if (token.IsEOF)
            return "EOF";

        string stepTypeText = string.Empty;
        string matchedItemsText = null;
        switch (token.MatchedType)
        {
            case TokenType.FeatureLine:
            case TokenType.ScenarioLine:
            case TokenType.ExamplesLine:
            case TokenType.DocStringSeparator:
            case TokenType.BackgroundLine:
            case TokenType.RuleLine:
                stepTypeText = "()";
                break;
            case TokenType.StepLine:
                var tokenType = token.MatchedGherkinDialect.GetStepKeywordType(token.MatchedKeyword);
                stepTypeText = $"({tokenType})";
                break;
            case TokenType.TagLine:
                matchedItemsText = string.Join(",", token.Line.GetTags().Select(i => i.Column + ":" + i.Text));
                break;
            case TokenType.TableRow:
                matchedItemsText = string.Join(",", token.Line.GetTableCells().Select(i => i.Column + ":" + i.Text));
                break;
        }

        return $"({token.Location.Line}:{token.Location.Column}){token.MatchedType}:{stepTypeText}{token.MatchedKeyword}/{token.MatchedText}/{matchedItemsText}";
    }
}

File: dart/lib/src/tokens/test_token_formatter.dart
import 'package:cucumber_gherkin/src/extensions/token_type_extension.dart';
import 'package:cucumber_gherkin/src/language/token.dart';

/// Renders a [Token] in the shared `*.tokens` text format used by the
/// acceptance tests.
class TestTokenFormatter {
  /// Returns the single-line `*.tokens` representation of [token].
  String formatToken(Token token) {
    if (token.isEof) {
      return 'EOF';
    }
    final items = token.matchedItems.map(
      (element) => '${element.column}:${element.text}',
    );
    final matchedItems = items.isEmpty ? '' : items.join(',');
    final matchedType = token.matchedType.wireName;
    final matchedKeyword =
        token.matchedKeyword.isEmpty ? '' : token.matchedKeyword;
    final matchedKeywordType =
        matchedKeyword.isEmpty
            ? ''
            : token.matchedKeywordType == null
            ? '()'
            : '(${token.matchedKeywordType!.value})';
    final matchedText = token.matchedText.isEmpty ? '' : token.matchedText;
    return '(${token.location.line}:${token.location.column})'
        '$matchedType:$matchedKeywordType$matchedKeyword/$matchedText/$matchedItems';
  }
}
